From 14d97093939c020c21d4166b2575c2f1cf75ff73 Mon Sep 17 00:00:00 2001
From: Ray Strode <rstrode@redhat.com>
Date: Thu, 19 Jun 2014 15:39:55 -0400
Subject: [PATCH] user-switch-applet: hide applet at start up if gdm not
 available

This commit checks if GDM is available at start up and hides the
item if necessary
---
 gui/user-switch-applet/applet.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/gui/user-switch-applet/applet.c b/gui/user-switch-applet/applet.c
index c31b3e3..a329ba6 100644
--- a/gui/user-switch-applet/applet.c
+++ b/gui/user-switch-applet/applet.c
@@ -6,60 +6,61 @@
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation; either version 2 of the License, or
  * (at your option) any later version.
  *
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU General Public License
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
  */
 
 #include "config.h"
 
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
 #include <sys/types.h>
 
 #include <glib/gi18n.h>
 #include <gdk/gdkkeysyms.h>
 #include <gtk/gtk.h>
 
 #include <gconf/gconf.h>
 #include <gconf/gconf-client.h>
 
 #include <dbus/dbus-glib.h>
+#include <dbus/dbus-glib-bindings.h>
 
 #include <bonobo/bonobo-main.h>
 #include <bonobo/bonobo-ui-util.h>
 
 #include <panel-applet.h>
 #include <panel-applet-gconf.h>
 
 #include "gdm-user-manager.h"
 #include "gdm-entry-menu-item.h"
 #include "gdm-settings-client.h"
 
 #define LOCKDOWN_DIR    "/desktop/gnome/lockdown"
 #define LOCKDOWN_USER_SWITCHING_KEY LOCKDOWN_DIR "/disable_user_switching"
 #define LOCKDOWN_LOCK_SCREEN_KEY    LOCKDOWN_DIR "/disable_lock_screen"
 #define LOCKDOWN_COMMAND_LINE_KEY   LOCKDOWN_DIR "/disable_command_line"
 
 typedef enum {
         GSM_PRESENCE_STATUS_AVAILABLE = 0,
         GSM_PRESENCE_STATUS_INVISIBLE,
         GSM_PRESENCE_STATUS_BUSY,
         GSM_PRESENCE_STATUS_IDLE,
 } GsmPresenceStatus;
 
 typedef struct _GdmAppletData
 {
         PanelApplet    *applet;
 
         GConfClient    *client;
         GdmUserManager *manager;
         GdmUser        *user;
@@ -682,69 +683,80 @@ maybe_lock_screen (GdmAppletData *adata)
 
         g_free (args[0]);
 }
 
 static void
 do_switch (GdmAppletData *adata,
            GdmUser       *user)
 {
         guint num_sessions;
 
         g_debug ("Do user switch");
 
         if (user == NULL) {
                 gdm_user_manager_goto_login_session (adata->manager);
                 goto out;
         }
 
         num_sessions = gdm_user_get_num_sessions (user);
         if (num_sessions > 0) {
                 gdm_user_manager_activate_user_session (adata->manager, user);
         } else {
                 gdm_user_manager_goto_login_session (adata->manager);
         }
  out:
         maybe_lock_screen (adata);
 }
 
 static void
 update_switch_user (GdmAppletData *adata)
 {
+        DBusGConnection *system_bus;
         gboolean can_switch;
         gboolean has_other_users;
+        gboolean has_gdm = FALSE;
+
+        system_bus = dbus_g_bus_get (DBUS_BUS_SYSTEM, NULL);
+        if (system_bus != NULL) {
+                DBusGProxy *bus_proxy;
+                bus_proxy = dbus_g_proxy_new_for_name (system_bus, DBUS_SERVICE_DBUS, DBUS_PATH_DBUS, DBUS_INTERFACE_DBUS);
+                org_freedesktop_DBus_name_has_owner (bus_proxy, "org.gnome.DisplayManager", &has_gdm, NULL);
+                g_object_unref (bus_proxy);
+                dbus_g_connection_unref (system_bus);
+        }
 
         can_switch = gdm_user_manager_can_switch (adata->manager);
         g_object_get (adata->manager,
                       "has-multiple-users", &has_other_users,
                       NULL);
 
-        if (can_switch && has_other_users) {
+        if (has_gdm && can_switch && has_other_users) {
                 gtk_widget_show (adata->login_screen_item);
         } else {
 
                 gtk_widget_hide (adata->login_screen_item);
         }
 }
 
 static void
 on_manager_is_loaded_changed (GdmUserManager *manager,
                               GParamSpec     *pspec,
                               GdmAppletData  *adata)
 {
         update_switch_user (adata);
 }
 
 static void
 on_manager_has_multiple_users_changed (GdmUserManager       *manager,
                                        GParamSpec           *pspec,
                                        GdmAppletData        *adata)
 {
         update_switch_user (adata);
 }
 
 #ifdef BUILD_PRESENSE_STUFF
 static void
 on_user_item_activate (GtkMenuItem   *item,
                        GdmAppletData *adata)
 {
         g_signal_stop_emission_by_name (item, "activate");
 }
-- 
1.9.3

